torch/testing/_internal/distributed/distributed_test.py

```diff
@@ -7513,3 +7513,48 @@ def test_ddp_bwd_with_retain_graph_static_single_fwd(self):
                 [param.grad for param in model_local.parameters()]
             )
             self.assertEqual(dist_grad_tensor, local_grad_tensor)
+
+        @unittest.skipIf(
+            BACKEND != "nccl" and BACKEND != "gloo",
+            "MPI backend does not support DDP communication hook on CUDA devices",
+        )
+        @skip_if_lt_x_gpu(2)
+        @skip_if_rocm
+        def test_hook_grad_idx_to_bucket_mapping(self):
+            # Test hook to ensure get_grad_index_to_variable_mapping works appropriately.
+            def _test_hook(hook_state, bucket: dist.GradBucket) -> torch.futures.Future:
+                # Run allreduce and wait for it to complete
+                allreduce_fut = default._allreduce_fut(None, bucket.get_tensor())
+                allreduce_fut.wait()
+                per_param_grad_tensors = bucket.get_per_parameter_tensors()
+                idx_to_model_params = bucket.get_grad_index_to_variable_mapping()
+                # Verify mapping is correct by ensuring param.grad is equal to
+                # the grad we get using the mapping.
+                m = {i : p.shape for i, p in idx_to_model_params.items()}
+                for i in range(len(per_param_grad_tensors)):
+                    grad = per_param_grad_tensors[i]
+                    model_param = idx_to_model_params[i]
+                    self.assertEqual(model_param.grad, grad)
```

Comment:
I think the question is: grad is averaged grad, model_param.grad is old grad, why are they equal here?

Reply:
It was an issue because the numerics were too small and tolerance was high enough to make them equal. Should be resolved now that we check for shape when grad_as_bucket_view=False.